Functions concept

bobtfish opened the pull request:
This is my first attempt at something as ambitious as a full exercise and a concept, so any/all feedback and tips very gratefully received.

Specific things that I'm unsure about / have questions on:
  * Fitting this concept/exercise into the existing hierarchy - I've done this somewhat, but if I add dependencies for everything that's relevant then I create cycles. Feedback welcomes from someone who has a more holistic view of the progression of concepts they think is right for the Go track.
  * The concept introduction and exercise introduction are nearly identical. Is this OK/appropriate? (Should I just be using .docs/introduction.md.tpl if so?)
  * .meta/design.md - this is my first attempt at one of these, I've read a few others but it's largely cargo cult.
  * Should my exercise include `forked_from`? It's not clear to me from the docs what the contents of this field would be? (I.e. is it just same track exercises, or can you reference cross tracks etc)

Fixes #1642

ErikSchierboom commented on exercises/concept/lasagna-master/.meta/config.json:
```diff
@@ -0,0 +1,12 @@
+    "test": ["lasagna-master_test.go"],
+    "exemplar": [".meta/exemplar.go"]
+  },
+  "source": "Inspired by the Lasagna Master exercise in the Javascript track",
+  "source_url": "https://github.com/exercism/javascript/blob/main/exercises/concept/lasagna-master/.docs/instructions.md"
```
You don't need the source if you specify the forked_from field (which is appropriate here)

```suggestion
  "forked_from": ["javascript/lasagna-master"]
```

ErikSchierboom commented on exercises/concept/lasagna-master/.meta/exemplar.go:
```diff
@@ -0,0 +1,35 @@
+package lasagna
+
+// Determines whether the lasagna is done based on the
+func cookingStatus(time int) string {
```
Should these functions start with a capital letter? I've seen that used in other exercises.

ErikSchierboom commented on exercises/concept/lasagna-master/lasagna-master.go:
```diff
@@ -0,0 +1,11 @@
+package lasagna
+
+/**
+ * Implement the functions needed to solve the exercise here.
+ * Here an example of the syntax as reminder:
+ *
+ * function yourFunction(param string, otherParam int) string {
+ *   ...
+ *   return "foo"
+ * }
+ */
```
I'd recommend doing the same here as in the [lasagna-exercise](https://github.com/exercism/go/blob/main/exercises/concept/lasagna/lasagna.go#L5), where there are TODOs for individual functions.

```suggestion
// TODO: define the 'CookingStatus()' function
```

ErikSchierboom commented on concepts/functions/about.md:
````diff
@@ -1,40 +1,167 @@
+}
+```
+
+## Pointers
+
+Functions in go pass their arguments by value.
+This means that the arguments are copied, and any changes to those arguments within the function will not be seen outside the function.
+
+```go
+val := 2
+func MultiplyByTwo(v int) int {
+    v = v * 2
+    return v
+}
+newval := MultiplyByTwo(val)
+// newval is 4, val is 2
+```
+
+To affect the values a function is called with, it is possible to pass pointer arguments, using the `*` syntax in the function, and the `&` syntax to generate a pointer.
+
+```go
+func swap(x, y *int) {
+    var temp int
+    temp = *x
+    *x = *y
+    *y = temp
+}
+
+var a int = 100
+var b int = 200
+swap(&a, &b)
+// a = 200, b = 100
+```
+
````
There will be a separate exercise on pointers: https://github.com/exercism/go/issues/1643

```suggestion
```

ErikSchierboom commented on concepts/functions/about.md:
```diff
@@ -1,40 +1,167 @@
-A function can have zero or more parameters. All parameters must be explicitly typed, there is no type inference for parameters. A function can also have multiple return values which must also be explicitly typed. Values are returned from functions using the [`return` keyword][return]. To allow a function, or a method to be [called by code in other packages][public-vs-private], the name of the fu ...
+## Return Values
+
+The function arguments are followed by zero or more return values which must also be explicitly typed.
+Single return values are left bare, multiple return values are wrapped in parenthesis.
+Values are returned to the calling code from functions using the [`return` keyword][return].
+There can be multiple `return` statements in a function.
+The execution of the function ends as soon as it hits one of those `return`s.
+If multiple values are to be returned from a function, they are comma seperated.
+More information about idiomatic use of [multiple return values][concept-multiple-return-values] can be found in the linked concept.
```
Nicely explained!

ErikSchierboom commented on concepts/functions/about.md:
```diff
@@ -1,40 +1,167 @@
-Functions in Go are considered [first class citizens][first-class-functions] making them very powerful. They can be used as:
+A function allows you to group code into a reusable unit.
+It consists of the `func` keyword, the name of the function, and a comma-separated list of zero or more parameters and types in round brackets.
+Other possibilities like [methods][concept-methods] and anonymous functions will be covered in other concepts.
```
I'm not a fan of referring to other concepts like this upfront. Maybe change these to be [concept widget links](https://exercism.org/docs/building/markdown/widgets) and move this text somewhat later in this document?

ErikSchierboom commented on concepts/functions/about.md:
````diff
@@ -1,40 +1,167 @@
-Invoking a function from inside the same package is done by specifying the function name and passing arguments for each of the function's parameters.
+## Variadic Functions
+
+Go does not in general support optional parameters, but does support [variadic functions][variadic-functions] which can capture zero or more arguments at the end of a function call.
+
+This is expressed by prefixing the type of the last argument with `...`, like this:
+
+```go
+func sum(nums ...int) int {
+    total := 0
+    for _, num := range nums {
+        total += num
+    }
+    return total
+}
+```
 
-Invoking a function from another package is done by specifying the package name as well as the function name.
+A similar syntax can be used to unpack a slice into a function call:
 
 ```go
-phrase := greeting.Hello("John")
+numbers := []int{1, 2, 4}
+result := sum(...numbers)
+// result => 7
````
Variadic functions are out of scope for this exercise (see the [issue description](https://github.com/exercism/go/issues/1642#issue-999186434)). Could you remove this section?

ErikSchierboom commented on concepts/functions/about.md:
````diff
@@ -1,40 +1,167 @@
+// a = 200, b = 100
+```
+
+## Public and Private Functions
+
+To allow a function to be called by [code in other packages][public-vs-private], the name of the function must start with a capital letter.
````
@junedev Do we want to include public/private here?

junedev replied:
No, that's not necessary. It is taught in the packages concept.

ErikSchierboom replied:
@bobtfish Could you remove the public/private section here (and in any other places in this PR)?

bobtfish replied:
This is not currently taught or mentioned anywhere in the packages concept or exercise that I can see, unless I'm missing it.

Does adding this need a ticket?

junedev replied:
Yes please, packages should teach how to import AND export things.

ErikSchierboom commented on concepts/functions/introduction.md:
````diff
@@ -1,46 +1,113 @@
-// hello is a private function
-func hello(name string) string {
-    return "Hello " + name
+Functions in go pass their arguments by value.
+This means that the arguments are copied, and any changes to those arguments within the function will not be seen outside the function.
+
+```go
+val := 2
+func MultiplyByTwo(v int) int {
+    v = v * 2
+    return v
 }
+newval := MultiplyByTwo(val)
+// newval is 4, val is 2
 ```
 
-Invoking a function is done by specifying the function name and passing arguments for each of the function's parameters.
+To affect the values a function is called with, it is possible to pass pointer arguments, using the `*` syntax in the function, and the `&` syntax to generate a pointer.
+
+```go
+func swap(x, y *int) {
+    var temp int
+    temp = *x
+    *x = *y
+    *y = temp
+}
 
-Go supports two types of comments. Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.
+var a int = 100
+var b int = 200
+swap(&a, &b)
+// a = 200, b = 100
+```
````
This section can be removed as this exercise should not cover pointers (nor is it needed by the student to solve the exercise).

bobtfish replied:
Damn, I knew I'd end up leaving something extraneous in the introduction, thanks for spotting!

ErikSchierboom commented on concepts/functions/introduction.md:
````diff
@@ -1,46 +1,113 @@
-```go
-package greeting
+As well as parameters, return values can optionally be named.
+If named return values are used, a `return` statement without arguments will return those values, this is known as a 'naked' return.
 
-// Hello is a public function
-func Hello(name string) string {
-    return hello(name)
+```go
+func SumAndMultiplyThenMinus(a, b, c int) (sum, mult int) {
+    sum, mult = a+b, a*b
+    sum -= c
+    mult -= c
+    return
 }
````
This example uses multiple returns, which is something that is introduced in the `card-tricks` exercise. The dependencies here are tricky.

bobtfish replied:
The other option is to return a `struct` type - the `structs` concept only depends on `numbers` and `conditonals-if` so this doesn't seem unreasonable...

bobtfish replied:
Per @junedev we're OK with multiple returns here, and will eventually remove them from `card-tricks` - resolving this.

ErikSchierboom reviewed (commented):
Thank you so much for doing this! I've roughly reviewed this, but decided to hold off on some reviewing more to first solve the dependency cycles.

> Fitting this concept/exercise into the existing hierarchy - I've done this somewhat, but if I add dependencies for everything that's relevant then I create cycles. Feedback welcomes from someone who has a more holistic view of the progression of concepts they think is right for the Go track.

@junedev could you perhaps take a look at how you envision the concept hierarchy? One immediate issue I could see was that this exercise depends on slices, but that the exercise teaching slices also teaches multiple returns. We could consider moving the multiple returns to this exercise? It's also important to note that we should decide if we want this to be an early exercise or not (as in: which exercises should depend on this)?

> The concept introduction and exercise introduction are nearly identical. Is this OK/appropriate? (Should I just be using .docs/introduction.md.tpl if so?)

This is perfectly ok. At some point the .tpl files will get good support, but we have only basic support for it now so I'd ignore it.

> .meta/design.md - this is my first attempt at one of these, I've read a few others but it's largely cargo cult.

It's perfectly fine.

> Should my exercise include forked_from? It's not clear to me from the docs what the contents of this field would be? (I.e. is it just same track exercises, or can you reference cross tracks etc)

Forked from is exclusively for cross-track references. So in this case it should be:

`"forked_from": ["javascript/lasagna-master"]

bobtfish commented:
> Thank you so much for doing this! I've roughly reviewed this, but decided to hold off on some reviewing more to first solve the dependency cycles.

Sounds good - thanks for the rough cut that was really helpful.

Agree that we should sort the high level dependency issues and decide where this fits and what it can/can't depend on before we get into the minutiae.

junedev commented:
Disclaimer: I did not read the PR content yet, I am mostly responding to the question regarding the concept tree.

The functions concept should include explaining multiple return values, not as a separate concept just as part of the content. This is where it belongs. Content can probably be copied. Thus multiple return values can be safely used as part of the exercise here.

This exercise should have slices and structs as prerequisites. When you modify the scaleRecipe exercise to use a slice of structs instead of a map that should be enough to make it work. (Make sure not to use `range`, it not available yet. Just use a normal for loop.)

We can then later go ahead and remove the multiple return values concept from slices. The exercise would just need a minimal change, there is no conceptual dependency (unlike for maps, there you need multiple return values). Until then I don't think anyone will think this is bad track because we repeat a couple of lines of concept content. Everything that depended on multiple return values will then depend on functions later.

The general hierarchy on the end would be that the student learns the basic types and conditionals, that unlocks slices and structs, which unlocks functions which unlocks maps.

With regards to pointers I would go with the "learn the details later approach". Strictly speaking you would need pointers to understand how the complex  types are passed to the function. However currently pointers has maps as prerequisite and I would like to keep it a bit further down the tree. So instead of going into the details here I would just talk about how the passing works for "reference types" and primitives.

Does that help/ make sense?

junedev commented:
I had a quick look at the exemplar file:
* I dont think the first task is needed here, it was about "not passing an argument" which does not exist in Go
* The second task should be changed to practice Gos version of a default parameter (an if statement). If the second argument is 0, use 2 as default.

ErikSchierboom commented:
@bobtfish Does @junedev's explanation help sort the dependencies issue?

bobtfish commented:
> @bobtfish Does @junedev's explanation help sort the dependencies issue?

Yes! Thanks @junedev - I'll work on getting this into shape in the next day or two and let you know when it's ready for review.

junedev commented:
Yes please, packages should teach how to import AND export things.
-------- Ursprüngliche Nachricht --------Von: Tomas Doran ***@***.***> Datum: 30.09.21  11:15  (GMT+01:00) An: exercism/go ***@***.***> Cc: June ***@***.***>, Mention ***@***.***> Betreff: Re: [exercism/go] Functions concept (#1720) 
@bobtfish commented on this pull request.



In concepts/functions/about.md:
> +## Public and Private Functions
+
+To allow a function to be called by [code in other packages][public-vs-private], the name of the function must start with a capital letter.

This is not currently taught or mentioned anywhere in the packages concept or exercise that I can see, unless I'm missing it.
Does adding this need a ticket?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

ErikSchierboom commented:
@junedev I've opened an issue: https://github.com/exercism/go/issues/1728

junedev commented:
@bobtfish Another thought regarding scaleRecipe. Instead of making it a slice of structs you could also just have a slice of floats an say "each item represents the amount of some ingredient ...". The task is about not modifying the argument so a slice will do to make this point.

junedev commented:
@bobtfish Setting this to draft for now. Mark as "ready for review" again whenever you are ready.

bobtfish commented:
> Make sure not to use `range`, it not available yet. Just use a normal for loop.

I'm unsure about this - it's totally possible to do this, but the exercise is a lot nicer to solve if you can use `range`, and there isn't anything in the dependencies at the moment which stops this from depending on `range-iteration`.

Why do we imagine that this has to be before we cover `range-iteration`?

junedev commented:
> Why do we imagine that this has to be before we cover range-iteration?

Range iteration need to show how to iterate over maps so maps need to be known to the student, maps need multiple return values from here so we can't move maps before functions.

bobtfish commented:
> > Why do we imagine that this has to be before we cover range-iteration?
> 
> Range iteration need to show how to iterate over maps so maps need to be known to the student, maps need multiple return values from here so we can't move maps before functions.

Aha, got it, thanks for the explanation. I'll take `range-iteration` out later tonight.

junedev commented on exercises/concept/lasagna-master/.docs/instructions.md:
```diff
@@ -0,0 +1,80 @@
+Implement a function `preparationTime` that accepts an array of layers as a `[]string` and the average preparation time per layer in minutes as an `int`.
+The function should return the estimate for the total preparation time based on the number of layers as an `int`.
+Go has no default values for functions.
+If the average preperation time is passed as `0` (the default value for an `int`), then the default value of `2` should be used.
```
```suggestion
If the average preperation time is passed as `0` (the default initial value for an `int`), then the default value of `2` should be used.
```

junedev commented on exercises/concept/lasagna-master/.meta/design.md:
```diff
@@ -0,0 +1,44 @@
+- `conditionals-if` is needed in the exercise
+- `range-iteration` is needed in the exercise
+- `slices` are needed in the exercise
+- `structs` are needed in the exercise
```
I think you removed this requirement.
```suggestion
```

junedev commented on exercises/concept/lasagna-master/.meta/exemplar.go:
```diff
@@ -0,0 +1,39 @@
+package lasagna
+
+// Estimates the preparation time based on the number of layers.
```
All function comments need to conform to the Go spec (see concept comments).

junedev reviewed (commented):
Well done overall! I really like the examples you used in the concepts.

Some more small things need to be fixed before this can be merged.

bobtfish commented:
> Well done overall! I really like the examples you used in the concepts.

Thanks!

> Some more small things need to be fixed before this can be merged and the merge conflict needs to be resolved.

All fixed up. I wasn't 100% sure what you were suggesting around the `PrintHelloName` example, but I have had a go at amending it.

I've also rebased and squashed as it was getting unwieldy - no point committing adding then removing then re-adding the same thing , and that also allowed me to fix the merge conflict so it'll merge cleanly without a load of back-and-forth merges to make future history exploration easy. (I'm not sure if there's a house style/recommendation about merges/rebases for PRs in progress)
